### PR TITLE
riscv32: LiteX VexRiscv support

### PR DESCRIPTION
**Pull request by mateusz-holenko**
This commit adds support for LiteX with VexRiscv softcore CPU, running
on Digilent ARTY development board, alongside with UART driver and
interrupt controller for LiteX and VexRiscv.

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Can you break this up into separate commits:

Each driver as one commit, the soc code as one commit, and the board code as one.

Thanks.

File: boards/riscv32/arty_litex_vexriscv/Kconfig.board
Comment (galak) [inline boards/riscv32/arty_litex_vexriscv/Kconfig.board]: The select HAS_DTS shouldn't be needed.  Should already be set for any RISC-V 32 port.

File: boards/riscv32/arty_litex_vexriscv/arty_litex_vexriscv.yaml
Comment (galak) [inline boards/riscv32/arty_litex_vexriscv/arty_litex_vexriscv.yaml]: Would be good to add the amount of memory / flash here.

File: drivers/interrupt_controller/vexriscv_litex.c
Comment (galak) [inline drivers/interrupt_controller/vexriscv_litex.c]: Is this copyright correct?

File: soc/riscv32/litex-vexriscv/dts_fixup.h
Comment (galak) [inline soc/riscv32/litex-vexriscv/dts_fixup.h]: We moved away from using CONFIG_ prefixed defines.  I'd suggest maybe seeing if you can just use the standard generated defines instead of needing a dts_fixup.h.

**Review by galak — CHANGES_REQUESTED** [review]
Some various clean up / questions.

File: drivers/interrupt_controller/vexriscv_litex.c
Comment (mateusz-holenko) [reply drivers/interrupt_controller/vexriscv_litex.c]: Do you mean year? The WIP version of this code was indeed published in 2018, so I'll update this.

File: boards/riscv32/arty_litex_vexriscv/doc/arty_litex_vexriscv.rst
Comment (dbkinder) [inline boards/riscv32/arty_litex_vexriscv/doc/arty_litex_vexriscv.rst]: First I've heard of this tool (flterm).  Does it do something that other serial port tools we typically mention don't do?

**Review by dbkinder — APPROVED** [review]
Docs look OK.  I'll let the developers comment on use of the flterm tools...

File: boards/riscv32/arty_litex_vexriscv/Kconfig.board
Comment (mateusz-holenko) [reply boards/riscv32/arty_litex_vexriscv/Kconfig.board]: You are right, we checked - it's not needed and will be removed.

File: soc/riscv32/litex-vexriscv/dts_fixup.h
Comment (mateusz-holenko) [reply soc/riscv32/litex-vexriscv/dts_fixup.h]: We'll use the generated defines directly.

File: boards/riscv32/arty_litex_vexriscv/doc/arty_litex_vexriscv.rst
Comment (pgielda) [reply boards/riscv32/arty_litex_vexriscv/doc/arty_litex_vexriscv.rst]: LiteX (SoC) has a built-in ROM bootloader ("bios") that supports uploading binaries but a special tool (e.g. flterm) is needed.

File: boards/riscv32/arty_litex_vexriscv/arty_litex_vexriscv.yaml
Comment (mateusz-holenko) [reply boards/riscv32/arty_litex_vexriscv/arty_litex_vexriscv.yaml]: There is no flash on the board, but we will add the ram size.

**Comment by enjoy-digital** [comment]
To ease support for others boards/SoCs in the future, i don't think the board directory should be 
`arty_litex_vexriscv`but maybe `litex-vexriscv` as you are doing for `soc/riscv32/litex-vexriscv/`.

Different LiteX generated SoCs can have the same exact parameters from Zephyr point of view, so it would be interesting to minimize information related to the boards. (only a few parameters will change between boards/SoCs).

**Comment by kgugala** [comment]
@enjoy-digital I agree, this could be a generic LiteX-VexRiscV port. The only thing arty dependant is RAM. This number is determined by the DDR chips on the board. Maybe we can reorganize the code so it adds generic LiteX support and only dts for Arty.

**Comment by enjoy-digital** [comment]
@kgugala: yes that seems fine.

**Comment by mateusz-holenko** [comment]
@galak I addressed all your questions, do you have any other comments?

**Comment by mateusz-holenko** [comment]
When working on @enjoy-digital's suggestion for reorganizing the code to add generic LiteX support, we realized it would be beneficial to move memory definitions from KConfig to DTS for riscv32 boards.

We will create a separate PR with that change and wait for it to be merged before carrying on with this one.

**Comment by mateusz-holenko** [comment]
A link to the above-mentioned PR: #15418.

File: boards/riscv32/arty_litex_vexriscv/Kconfig.board
Comment (mateusz-holenko) [reply boards/riscv32/arty_litex_vexriscv/Kconfig.board]: It's done.

File: drivers/interrupt_controller/vexriscv_litex.c
Comment (mateusz-holenko) [reply drivers/interrupt_controller/vexriscv_litex.c]: I changed copyright year to 2018-2019 in every file.

File: soc/riscv32/litex-vexriscv/dts_fixup.h
Comment (mateusz-holenko) [reply soc/riscv32/litex-vexriscv/dts_fixup.h]: It's done.

**Comment by mateusz-holenko** [comment]
Since the #15418 is merged to master, we can proceed with this PR.

@enjoy-digital As you suggested, the platform has been renamed to `litex_vexriscv`. There is a separate dts file that includes SoC and adds memory definition, but we decided to drop the reference to Arty board making the platform a generic LiteX.

@galak I have rebased it onto the current master, addressed all your comments. Are we ready to merge it?

**Comment by enjoy-digital** [comment]
@mateusz-holenko: great, thanks.

File: soc/riscv32/litex-vexriscv/Kconfig.defconfig
Comment (galak) [inline soc/riscv32/litex-vexriscv/Kconfig.defconfig]: I assume from the comment you are saying that interrupt's aren't supported.  Is this really accomplishing what you want?

File: soc/riscv32/litex-vexriscv/soc.h
Comment (galak) [inline soc/riscv32/litex-vexriscv/soc.h]: Is this define need anywhere but the IRQ code?  If not, just move it there.

**Review by galak — CHANGES_REQUESTED** [review]
Some minor cleanups.

File: dts/riscv32/riscv32-litex-vexriscv.dtsi
Comment (mateusz-holenko) [reply dts/riscv32/riscv32-litex-vexriscv.dtsi]: I set proper names in `reg-names` and referenced to them directly in the driver.

File: dts/riscv32/riscv32-litex-vexriscv.dtsi
Comment (mateusz-holenko) [reply dts/riscv32/riscv32-litex-vexriscv.dtsi]: I changed the name to be more specific. VexRiscv is an independent core that can be used outside LiteX, so I set compatible to `spinalhdl,vexriscv` and left the second value `riscv` as I saw this is the pattern for other RISC-V cores.

File: soc/riscv32/litex-vexriscv/Kconfig.defconfig
Comment (mateusz-holenko) [reply soc/riscv32/litex-vexriscv/Kconfig.defconfig]: You are right, that was done wrong. I removed this line from soc-level Kconfig and added `depends on !SERIAL_SUPPORT_INTERRUPT` in uart-level Kconfig.

File: soc/riscv32/litex-vexriscv/soc.h
Comment (mateusz-holenko) [reply soc/riscv32/litex-vexriscv/soc.h]: I got rid of this define at all.

**Comment by mateusz-holenko** [comment]
@galak I addressed all comments.

File: CODEOWNERS
Comment (galak) [inline CODEOWNERS]: Can you change this to be:
```
/dts/bindings/*/vexrisc*
/dts/bindings/*/litex*
```

there appears to be some bug if we have `,` in this file.

**Review by galak — CHANGES_REQUESTED** [review]
need a minor change in CODEOWNERS due to a github bug.

File: CODEOWNERS
Comment (mateusz-holenko) [reply CODEOWNERS]: @galak I refactored CODEOWNERS as suggested.

**Review by galak — APPROVED** [review]

**Comment by mgielda** [comment]
OK, @galak can you merge?

**Pull request merged**
